python/pyspark/sql/connect/session.py

```diff
@@ -215,7 +215,31 @@ def createDataFrame(
         _inferred_schema: Optional[StructType] = None
 
         if isinstance(data, pd.DataFrame):
-            _table = pa.Table.from_pandas(data)
+            from pandas.api.types import (  # type: ignore[attr-defined]
+                is_datetime64_dtype,
+                is_datetime64tz_dtype,
+            )
+
+            # We need double conversions for the truncation, first truncate to microseconds.
+            for col in data:
+                print("Checking", col)
+                if is_datetime64tz_dtype(data[col].dtype):
+                    data[col] = data[col].astype("datetime64[us, UTC]")
+                elif is_datetime64_dtype(data[col].dtype):
+                    data[col] = data[col].astype("datetime64[us]")
+
+            # Create a new schema and change the types to the truncated microseconds.
+            pd_schema = pa.Schema.from_pandas(data)
```

Comment:
just for my understanding, why we need to create the pyarrow schema here?   the Pandas DataFrame datatypes had already been modified above.

Reply:
Even modifying the schema, the Pandas -> Arrow conversion complains that the type is nanoseconds. The pandas operation allows the truncation to actually proceed.

If I skip the pandas operation the error is:

```
pyarrow.lib.ArrowInvalid: ('Casting from timestamp[ns] to timestamp[us] would lose data: 1546329600000000500', 'Conversion failed for column aware with type datetime64[ns, UTC]')
```

If I skip the second one the error is:

```
pyspark.sql.connect.client.SparkConnectException: (org.apache.spark.SparkUnsupportedOperationException) Unsupported data type: Timestamp(NANOSECOND, null)
```